### PR TITLE
Added task definition and configs for Panoptic Mask R-CNN 

### PR DESCRIPTION
**Pull request by srihari-humbarwadi**
# Description
Added task definition for Panoptic Mask R-CNN 

## Type of change
- [x] A new research paper code implementation

## Tests
 - `PanopticMaskRCNNTask`
   - `python -m official.vision.beta.projects.panoptic_maskrcnn.tasks.panoptic_maskrcnn_test`
 - configs
   - `python -m official.vision.beta.projects.panoptic_maskrcnn.configs.panoptic_maskrcnn_test`

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py [hunk 1]
```diff
@@ -13,35 +13,128 @@
...
 
 import dataclasses
+
+from official.core import config_definitions as cfg
+from official.core import exp_factory
+from official.modeling import optimization
 from official.vision.beta.configs import maskrcnn
 from official.vision.beta.configs import semantic_segmentation
 
+# pylint: disable=missing-class-docstring
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:27]: remove and add class docstring.

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py [hunk 2]
```diff
@@ -13,35 +13,128 @@
...
 
 @dataclasses.dataclass
 class Parser(maskrcnn.Parser):
-  """Panoptic Segmentation parser config."""
-  # If resize_eval_groundtruth is set to False, original image sizes are used
-  # for eval. In that case, groundtruth_padded_size has to be specified too to
-  # allow for batching the variable input sizes of images.
-  resize_eval_segmentation_groundtruth: bool = True
+  # If segmentation_resize_eval_groundtruth is set to False, original image
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:31]: revert these changes.

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py [hunk 3]
```diff
@@ -13,35 +13,128 @@
...
   """Panoptic Mask R-CNN model config."""
   segmentation_model: semantic_segmentation.SemanticSegmentationModel = (
       semantic_segmentation.SemanticSegmentationModel(num_classes=2))
+  include_mask = True
   shared_backbone: bool = True
   shared_decoder: bool = True
+
+@dataclasses.dataclass
+class Losses(maskrcnn.Losses):
+  segmentation_loss: semantic_segmentation.Losses = \
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:56]: Can you flatten this?

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py [hunk 4]
```diff
@@ -13,35 +13,128 @@
...
+      semantic_segmentation.Losses()
+  segmentation_weight: float = 0.5
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:58]: semantic_segmentation_weight, also can you default to 1.0?

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py [hunk 5]
```diff
@@ -13,35 +13,128 @@
...
+  init_checkpoint_modules: Optional[List[str]] = dataclasses.field(
+      default_factory=list)
+
+COCO_INPUT_PATH_BASE = 'coco'
+
+
+@exp_factory.register_config_factory('panoptic_maskrcnn_resnetfpn_coco')
+def panoptic_maskrcnn_resnetfpn_coco() -> cfg.ExperimentConfig:
+  """COCO panoptic segmentation with Panoptic Mask R-CNN."""
+  steps_per_epoch = 500
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:81]: Can you compute that from coco training samples, and train_batch_size?

File: official/vision/beta/projects/panoptic_maskrcnn/tasks/panoptic_maskrcnn.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+from official.common import dataset_fn
+from official.core import task_factory
+from official.vision.beta.dataloaders import input_reader_factory
+from official.vision.beta.dataloaders import tf_example_label_map_decoder
+from official.vision.beta.evaluation import coco_evaluator
+from official.vision.beta.evaluation import segmentation_metrics
+from official.vision.beta.losses import segmentation_losses
+from official.vision.beta.tasks import maskrcnn
+from official.vision.beta.projects.panoptic_maskrcnn.modeling import factory
+from official.vision.beta.projects.panoptic_maskrcnn.configs \
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/tasks/panoptic_maskrcnn.py:29]: remove \ from imports.

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py [hunk 7]
```diff
@@ -13,35 +13,128 @@
...
+class PanopticMaskRCNNTask(maskrcnn.MaskRCNNTask):
+  model: PanopticMaskRCNN = PanopticMaskRCNN()
+  train_data: DataConfig = DataConfig(is_training=True)
+  validation_data: DataConfig = DataConfig(is_training=False,
+                                           drop_remainder=False)
+  segmentation_evaluation: semantic_segmentation.Evaluation = \
+      semantic_segmentation.Evaluation()
+  losses: Losses = Losses()
+  backbone_init_checkpoint: Optional[str] = None
+  segmentation_backbone_init_checkpoint: Optional[str] = None
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:71]: segmentation_init_checkpoint.

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py [hunk 8]
```diff
@@ -13,35 +13,128 @@
...
+@dataclasses.dataclass
+class PanopticMaskRCNNTask(maskrcnn.MaskRCNNTask):
+  model: PanopticMaskRCNN = PanopticMaskRCNN()
+  train_data: DataConfig = DataConfig(is_training=True)
+  validation_data: DataConfig = DataConfig(is_training=False,
+                                           drop_remainder=False)
+  segmentation_evaluation: semantic_segmentation.Evaluation = \
+      semantic_segmentation.Evaluation()
+  losses: Losses = Losses()
+  backbone_init_checkpoint: Optional[str] = None
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:70]: init_checkpoint

File: official/vision/beta/projects/panoptic_maskrcnn/tasks/panoptic_maskrcnn.py [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+        status.assert_consumed()
+
+      elif init_module == 'backbone':
+        checkpoint_path = _get_checkpoint_path(
+            self.task_config.backbone_init_checkpoint)
+        ckpt = tf.train.Checkpoint(backbone=model.backbone)
+        status = ckpt.restore(checkpoint_path)
+        status.expect_partial().assert_existing_objects_matched()
+
+      elif init_module == 'segmentation_backbone':
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/tasks/panoptic_maskrcnn.py:91]: can you add segmentation_decoder as well?

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py [hunk 10]
```diff
@@ -13,35 +13,128 @@
...
+  segmentation_backbone_init_checkpoint: Optional[str] = None
+  init_checkpoint_modules: Optional[List[str]] = dataclasses.field(
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:72]: Can you list here the available options?

File: official/vision/beta/projects/panoptic_maskrcnn/tasks/panoptic_maskrcnn.py [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+        metrics.append(tf.keras.metrics.Mean(name, dtype=tf.float32))
+      
+      if self.task_config.segmentation_evaluation.report_train_mean_iou:
+        self.segmentation_train_mean_iou = segmentation_metrics.MeanIoU(
+            name='train_mean_iou',
+            num_classes=self.task_config.model.num_classes,
+            rescale_predictions=False,
+            dtype=tf.float32)
+
+    else:
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/tasks/panoptic_maskrcnn.py:223]: Why do you have else statement here? can we always have coco metric?

Comment (srihari-humbarwadi) [reply official/vision/beta/projects/panoptic_maskrcnn/tasks/panoptic_maskrcnn.py:223]: For evaluating instance masks, `COCOEvaluator` expects an annotation_file since the eval dataloader does not return instance mask groundtruths. Also, the task config supports only a single annotation_file, so we run COCOEvaluator only when `training=False`

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py (hunk 2 above)
Comment (arashwan) [reply official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:31]: Oh, no I think that's fine, just add back the docstring.

File: official/vision/beta/projects/panoptic_maskrcnn/tasks/panoptic_maskrcnn.py (hunk 11 above)
Comment (arashwan) [reply official/vision/beta/projects/panoptic_maskrcnn/tasks/panoptic_maskrcnn.py:223]: Oh sorry, no keep it as is. I thought this is the else statement for "if self.task_config.segmentation_evaluation.report_train_mean_iou:"

File: official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py [hunk 12]
```diff
@@ -13,35 +13,139 @@
...
   segmentation_model: semantic_segmentation.SemanticSegmentationModel = (
       semantic_segmentation.SemanticSegmentationModel(num_classes=2))
+  include_mask = True
   shared_backbone: bool = True
   shared_decoder: bool = True
+
+@dataclasses.dataclass
+class Losses(maskrcnn.Losses):
+  """Panoptic Mask R-CNN loss config."""
+  segmentation_loss: semantic_segmentation.Losses = semantic_segmentation.Losses()  # pylint: disable=line-too-long
```
Comment (arashwan) [inline official/vision/beta/projects/panoptic_maskrcnn/configs/panoptic_maskrcnn.py:57]: Can you flatten this in the since that the losses don't have any hierarchy?

**Review by arashwan — APPROVED** [review]

**Review by jaeyounkim — APPROVED** [review]
